File: app/src/main/kotlin/com/hermesandroid/relay/data/Endpoint.kt
package com.hermesandroid.relay.data

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.net.URI

/**
 * One entry in a pairing payload's `endpoints` array (ADR 24 — multi-endpoint
 * pairing, 2026-04-19).
 *
 * Pairing QRs can now carry an ordered list of candidate endpoints so a single
 * pairing works across LAN / Tailscale / public-reverse-proxy networks. The
 * phone picks the highest-priority reachable candidate at connect time and
 * re-evaluates on network change.
 *
 * Wire contract (v3 pairing payload):
 * ```json
 * {
 *   "role": "lan",
 *   "priority": 0,
 *   "api":   { "host": "192.168.1.100", "port": 8642, "tls": false },
 *   "relay": { "url": "ws://192.168.1.100:8767", "transport_hint": "ws" }
 * }
 * ```
 *
 * **Semantics (locked by ADR 24):**
 *  - [role] is an open string. Known values `lan` / `tailscale` / `public`
 *    get styled labels; anything else renders generically (`Custom route (<role>)`).
 *    No enum, no normalization — the raw role string must round-trip exactly
 *    so HMAC canonicalization holds.
 *  - [priority] is strict, `0 = highest`. Reachability never promotes a lower
 *    priority over a higher one; it only breaks ties among equal priorities.
 *  - The per-endpoint [RelayEndpoint] intentionally carries **only** the URL
 *    and transport hint. The pairing `code`, `ttl_seconds`, and `grants` stay
 *    at the top level of the pairing payload because they're per-pair
 *    artifacts — not per-endpoint.
 */
@Serializable
data class EndpointCandidate(
    val role: String,
    val priority: Int = 0,
    /** Optional legacy/API-server surface. Dashboard-only routes omit it. */
    val api: ApiEndpoint? = null,
    /** Optional Hermes-Relay bridge surface. Standard upstream routes omit it. */
    val relay: RelayEndpoint? = null,
    val dashboard: DashboardEndpoint? = null,
    val proxy: ProxyEndpoint? = null,
    /** Optional outbound rendezvous carrying the pinned [proxy] byte stream. */
    val broker: BrokerEndpoint? = null,
    val security: String? = null,
    val recommended: Boolean = false,
    val experimental: Boolean = false,
    @SerialName("display_name")
    val displayName: String? = null,
)

/**
 * The API-server half of an [EndpointCandidate] — the HTTP/SSE target the
 * phone uses for `/v1/runs`, `/v1/chat/completions`, `/api/sessions/...`, etc.
 *
 * Note: [tls] defaults to false so a v2-synthesized candidate (built from a
 * legacy QR with no `endpoints` field and no top-level `tls`) still
 * deserializes cleanly.
 */
@Serializable
data class ApiEndpoint(
    val host: String,
    val port: Int,
    val tls: Boolean = false,
) {
    /** Build the full API server URL from host, port, and tls flag. */
    val url: String
        get() = "${if (tls) "https" else "http"}://$host:$port"
}

/**
 * Dashboard/admin surface for an [EndpointCandidate]. This is optional so
 * older v3 payloads that only carried API + Relay endpoints keep
 * deserializing; when absent, Android derives the conventional same-host
 * `:9119` dashboard URL from [ApiEndpoint].
 */
@Serializable
data class DashboardEndpoint(
    val url: String,
)

/**
 * The relay-server half of an [EndpointCandidate] — the WSS URL the phone
 * opens for the bridge + terminal channels.
 *
 * @property url full WebSocket URL, e.g. `ws://192.168.1.100:8767` (dev) or
 *           `wss://hermes.example.com/relay` (fronted by a reverse proxy).
 * @property transportHint `"wss"` / `"ws"` / `null`. Drives the plaintext-ws
 *           consent gate and the transport-security UI badge. Never gates
 *           behavior on its own — the scheme of [url] is authoritative.
 */
@Serializable
data class RelayEndpoint(
    val url: String,
    @SerialName("transport_hint")
    val transportHint: String? = null,
)

/**
 * Optional plugin-owned secure proxy route. Unlike [api], [dashboard], and
 * [relay], this is one app-facing base that can cover all Hermes-Relay
 * supported traffic after pairing. It is deliberately optional so plugin
 * proxy support can be advertised by newer payloads without changing the
 * standard upstream connection model.
 */
@Serializable
data class ProxyEndpoint(
    val url: String,
    @SerialName("transport_hint")
    val transportHint: String? = null,
    @SerialName("pin_sha256")
    val pinSha256: String? = null,
    /** Independently authenticated services carried by this pinned origin. */
    val surfaces: List<String> = listOf("relay"),
)

/**
 * Hermes Reach rendezvous metadata from an operator-reviewed pairing payload.
 * The token authenticates only this broker route; Hermes service credentials
 * remain inside the QR-pinned Secure Link TLS connection.
 */
@Serializable
data class BrokerEndpoint(
    val url: String,
    @SerialName("protocol_version")
    val protocolVersion: Int = 1,
    @SerialName("host_id")
    val hostId: String,
    @SerialName("credential_kind")
    val credentialKind: String,
    val token: String,
    @SerialName("expires_at")
    val expiresAt: Long? = null,
)

/**
 * Returns true when [EndpointCandidate.role] is one of the built-in, styled
 * roles: `lan`, `tailscale`, or `public`. Case-insensitive match — but the
 * role string itself is still preserved verbatim for HMAC canonicalization.
 *
 * Unknown roles (`"wireguard"`, `"zerotier"`, `"netbird-eu"`, operator-defined
 * labels) return false so the UI can fall back to [displayLabel]'s generic
 * "Custom route" treatment.
 */
fun EndpointCandidate.isKnownRole(): Boolean {
    return when (role.lowercase()) {
        "lan", "tailscale", "public", "plugin_proxy", "plugin-proxy", "outbound_broker", "https",
        "dashboard", "authenticated_dashboard" -> true
        else -> false
    }
}

/**
 * Human-readable label for the UI. Known roles get fixed-case styled labels;
 * unknown roles render as `"Custom route (<role>)"` with the raw role preserved
 * so an operator can see exactly what they labeled it.
 *
 * The raw [role] on the [EndpointCandidate] is NOT modified — it stays in its
 * emitted form for HMAC canonicalization. This is a display-only transform.
 */
fun EndpointCandidate.displayLabel(): String {
    return when (role.lowercase()) {
        "lan" -> "LAN"
        "tailscale" -> "Tailscale"
        "public" -> if (primaryRouteUrl()?.startsWith("https://", ignoreCase = true) == true) {
            "HTTPS"
        } else {
            "Public"
        }
        "https" -> "HTTPS"
        "dashboard", "authenticated_dashboard" -> if (
            primaryRouteUrl()?.startsWith("https://", ignoreCase = true) == true
        ) {
            "HTTPS Dashboard"
        } else {
            "Dashboard"
        }
        "plugin_proxy", "plugin-proxy" -> "Hermes Secure Link"
        "outbound_broker", "broker", "relay_broker" -> "Hermes Reach · Experimental"
        else -> displayName?.trim()?.takeIf { it.isNotBlank() } ?: "Custom route ($role)"
    }
}

/**
 * True for a route created only to keep Dashboard/Gateway authentication on
 * its canonical origin. It is a service address, not another selectable
 * whole-connection or VPN route.
 */
fun EndpointCandidate.isDashboardOnlyRoute(): Boolean =
    api == null && relay == null && proxy == null && broker == null && dashboard != null

/** Dashboard-first URL identity for routing, diagnostics, and UI labels. */
fun EndpointCandidate.primaryRouteUrl(): String? =
    dashboard?.url?.trim()?.trimEnd('/')?.takeIf { it.isNotBlank() }
        ?: api?.url
        ?: relay?.url?.trim()?.trimEnd('/')?.takeIf { it.isNotBlank() }
        ?: proxy?.url?.trim()?.trimEnd('/')?.takeIf { it.isNotBlank() }

/** Dashboard/Gateway identity only; Relay and broker transports are extensions. */
fun EndpointCandidate.gatewayRouteUrl(): String? =
    dashboard?.url?.trim()?.trimEnd('/')?.takeIf { it.isNotBlank() }
        ?: api?.url?.let(Connection::deriveDefaultDashboardUrl)

/** Stable host/port identity without assuming that an API surface exists. */
fun EndpointCandidate.routeAuthority(): String? {
    val rawUrl = primaryRouteUrl() ?: return null
    val httpUrl = when {
        rawUrl.startsWith("ws://", ignoreCase = true) -> "http://${rawUrl.substringAfter("://")}"
        rawUrl.startsWith("wss://", ignoreCase = true) -> "https://${rawUrl.substringAfter("://")}"
        else -> rawUrl
    }
    val uri = runCatching { URI(httpUrl) }.getOrNull() ?: return null
    val host = uri.host?.lowercase()?.takeIf { it.isNotBlank() } ?: return null
    val port = when {
        uri.port > 0 -> uri.port
        uri.scheme.equals("https", ignoreCase = true) -> 443
        else -> 80
    }
    return "$host:$port"
}

fun EndpointCandidate.hasSecureProxy(): Boolean =
    proxy?.isValidPinnedProxy() == true

/** Product-facing service inventory; wire identifiers remain unchanged. */
fun EndpointCandidate.secureLinkServices(): List<String> =
    if (!hasSecureProxy()) emptyList() else proxy.orEmptySurfaces()

fun EndpointCandidate.secureLinkCoversAllServices(): Boolean =
    secureLinkServices().containsAll(listOf("relay", "api", "dashboard"))

fun EndpointCandidate.presentationRouteUrl(): String? =
    broker?.url?.takeIf { hasHermesReach() } ?: proxy?.url?.takeIf { hasSecureProxy() } ?: primaryRouteUrl()

fun EndpointCandidate.hasHermesReach(): Boolean =
    role.lowercase() in setOf("outbound_broker", "broker", "relay_broker") &&
        broker?.isValidHermesReach() == true && hasSecureProxy()

fun BrokerEndpoint.isValidHermesReach(): Boolean {
    if (protocolVersion != 1 || !hostId.isCanonicalBase64Url(16) || !token.isCanonicalBase64Url(32)) return false
    if (credentialKind !in setOf("bootstrap", "route")) return false
    if (credentialKind == "bootstrap" && expiresAt?.let { it <= System.currentTimeMillis() / 1000L } == true) return false
    val uri = runCatching { URI(url.trim()) }.getOrNull() ?: return false
    if (!uri.scheme.equals("wss", ignoreCase = true) || uri.host.isNullOrBlank()) return false
    if (!uri.rawUserInfo.isNullOrBlank() || uri.rawQuery != null || uri.rawFragment != null) return false
    return uri.rawPath.orEmpty().let { it.isEmpty() || it == "/" || it == "/v1/connect" }
}

private fun String.isCanonicalBase64Url(byteCount: Int): Boolean {
    if (isBlank() || '=' in this) return false
    val decoded = runCatching { java.util.Base64.getUrlDecoder().decode(this) }.getOrNull() ?: return false
    return decoded.size == byteCount &&
        java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(decoded) == this
}

/** Exact host locator + broker authority replacement; never crosses devices. */
internal fun replaceHermesReachCredential(
    source: List<EndpointCandidate>,
    expected: BrokerEndpoint,
    replacement: EndpointCandidate,
): List<EndpointCandidate> = source.map { candidate ->
    if (candidate.broker?.hostId == expected.hostId &&
        sameBrokerAuthority(candidate.broker.url, expected.url)
    ) replacement else candidate
}

internal fun sameBrokerAuthority(left: String, right: String): Boolean = runCatching {
    val a = URI(left.trim())
    val b = URI(right.trim())
    fun port(uri: URI) = if (uri.port > 0) uri.port else 443
    a.scheme.equals("wss", true) && b.scheme.equals("wss", true) &&
        a.host.equals(b.host, true) && port(a) == port(b) &&
        a.rawPath.orEmpty().trimEnd('/') == b.rawPath.orEmpty().trimEnd('/')
}.getOrDefault(false)

private fun ProxyEndpoint?.orEmptySurfaces(): List<String> = this?.surfaces.orEmpty()
    .map { it.trim().lowercase() }
    .filter { it in setOf("relay", "api", "dashboard") }
    .distinct()

fun ProxyEndpoint.isValidPinnedProxy(): Boolean {
    val uri = runCatching { URI(url.trim().trimEnd('/')) }.getOrNull() ?: return false
    if (!uri.scheme.equals("https", ignoreCase = true) || uri.host.isNullOrBlank()) return false
    if (!uri.rawUserInfo.isNullOrBlank() || uri.rawQuery != null || uri.rawFragment != null) return false
    if (uri.rawPath.orEmpty().let { it.isNotEmpty() && it != "/" }) return false
    val pin = pinSha256?.trim()?.removePrefix("sha256/") ?: return false
    return runCatching { java.util.Base64.getDecoder().decode(pin).size == 32 }.getOrDefault(false)
}

File: desktop/src/endpoint.ts
// Multi-endpoint pairing primitives — pure types + tiny helpers (ADR 24).
//
// Mirrors `app/src/main/kotlin/com/hermesandroid/relay/data/Endpoint.kt`.
// The Kotlin data classes over there are the source of truth — any semantic
// change must go through ADR 24 and land in both surfaces simultaneously.
//
// Intentionally pure: no I/O, no Node globals beyond TS. Probe semantics +
// the parser live in `pairingQr.ts`; this file is safe to import from
// anywhere, including tests that have no network access.

/**
 * Known endpoint roles. The wire format treats `role` as an open string —
 * operators can emit arbitrary labels — so anything not in this set
 * normalizes to `'custom'` for display. The **raw** string is preserved
 * verbatim in `EndpointCandidate.role` for HMAC canonicalization; this
 * narrowed type is display-only.
 */
export type EndpointRole = 'lan' | 'tailscale' | 'public' | 'custom'

/**
 * API-server half of an endpoint candidate. Points at the HTTP/SSE target
 * used for `/v1/runs`, `/v1/chat/completions`, `/api/sessions/*`, etc.
 *
 * `tls` defaults to `false` on the wire so a legacy v1/v2 payload without
 * the field synthesizes cleanly (see `payloadToCandidates`).
 */
export interface ApiEndpoint {
  host: string
  port: number
  tls: boolean
}

/** Dashboard/Gateway half of a pairing candidate. */
export interface DashboardEndpoint {
  url: string
}

/** Build the full API URL — mirrors Kotlin's `ApiEndpoint.url` getter. */
export function apiUrl(e: ApiEndpoint): string {
  return `${e.tls ? 'https' : 'http'}://${e.host}:${e.port}`
}

/**
 * Relay-server half of an endpoint candidate. Carries only the URL and the
 * transport hint — the pairing `code`, `ttl_seconds`, and `grants` are
 * per-pair artifacts that stay on the top-level pairing payload.
 */
export interface RelayEndpoint {
  url: string
  /** `"wss"` | `"ws"` | undefined. UI hint only — `url` scheme is authoritative. */
  transportHint?: string
}

/**
 * One entry in a v3 pairing payload's `endpoints` array. `priority` is
 * strict — `0 = highest`. Reachability is a tiebreaker **within** a
 * priority tier; it never promotes a lower tier over a higher one.
 */
export interface EndpointCandidate {
  /** Raw role string from the wire — preserved verbatim for HMAC canonicalization. */
  role: string
  priority: number
  /** Optional legacy API-server surface. Dashboard-first routes omit it. */
  api?: ApiEndpoint
  relay: RelayEndpoint
  /** Optional standard Dashboard/Gateway origin. */
  dashboard?: DashboardEndpoint
  /** Native plugin secure-proxy advertisement. Preserved even when the
   * desktop currently consumes only its relay WSS path. */
  proxy?: {
    url: string
    transportHint?: string
    pinSha256: string
    certificateDerBase64?: string
    surfaces?: string[]
  }
  security?: string
  recommended?: boolean
  /** Experimental routes are supported for explicit evaluation but are
   * never presented as the default remote-access recommendation. */
  experimental?: boolean
  broker?: {
    url: string
    hostId: string
    credentialKind: 'bootstrap' | 'route'
    token: string
    expiresAt?: string | number | null
  }
}

export function candidateDisplayLabel(candidate: Pick<EndpointCandidate, 'role' | 'broker'> & Partial<Pick<EndpointCandidate, 'relay'>>): string {
  const role = candidate.role.toLowerCase()
  if (candidate.broker && (role === 'outbound_broker' || role === 'broker' || role === 'relay_broker')) return 'Hermes Reach (experimental)'
  // Older session files used the generic `custom` role even for ordinary
  // private-address routes. Keep named operator-defined roles intact, but
  // classify that legacy placeholder from its actual relay URL.
  if (role === 'custom' && candidate.relay?.url) return displayLabel(inferEndpointRole(candidate.relay.url))
  return displayLabel(candidate.role)
}

/**
 * Narrow a raw role string to one of the known UI roles. Case-insensitive.
 * Anything outside the built-in set (operator-defined roles like
 * `"wireguard"`, `"zerotier"`, `"netbird-eu"`) folds to `'custom'` so
 * callers can switch on a closed set.
 *
 * This is a display-time transform. The raw role on the candidate stays
 * in its emitted form — do NOT mutate `EndpointCandidate.role` based on
 * this narrowing or the HMAC canonicalization breaks.
 */
export function parseRawRole(raw: string): EndpointRole {
  switch (raw.toLowerCase()) {
    case 'lan':
      return 'lan'
    case 'tailscale':
      return 'tailscale'
    case 'public':
      return 'public'
    default:
      return 'custom'
  }
}

/**
 * True when `role` maps to a built-in styled role. Mirrors Kotlin's
 * `EndpointCandidate.isKnownRole()`.
 */
export function isKnownRole(role: string): boolean {
  return role.toLowerCase() === 'https' || parseRawRole(role) !== 'custom'
}

/**
 * Human-readable label for the role. Mirrors Kotlin's `displayLabel()`.
 * Unknown roles render as `"Custom VPN (<raw>)"` so the operator sees
 * exactly what they labeled it.
 */
export function displayLabel(role: string): string {
  const normalized = role.toLowerCase()
  if (normalized === 'https') return 'HTTPS'
  if (normalized === 'plugin_proxy') return 'Hermes Secure Link'
  if (normalized === 'outbound_broker' || normalized === 'relay_broker' || normalized === 'broker') return 'Hermes Reach (experimental)'
  switch (parseRawRole(role)) {
    case 'lan':
      return 'LAN'
    case 'tailscale':
      return 'Tailscale'
    case 'public':
      return 'Public'
    default:
      return `Custom VPN (${role})`
  }
}

/** Infer a built-in role for legacy/direct sessions that did not persist one. */
export function inferEndpointRole(rawUrl: string): EndpointRole {
  try {
    const parsed = new URL(rawUrl)
    const host = parsed.hostname.replace(/^\[|\]$/g, '').toLowerCase()
    const octets = host.split('.').map(Number)
    const ipv4 = octets.length === 4 && octets.every(part => Number.isInteger(part) && part >= 0 && part <= 255)

    if (host.endsWith('.ts.net') || (ipv4 && octets[0] === 100 && octets[1] >= 64 && octets[1] <= 127)) {
      return 'tailscale'
    }
    if (
      host === 'localhost' || host.endsWith('.local') || host === '::1' ||
      host.startsWith('fc') || host.startsWith('fd') ||
      /^(fe8|fe9|fea|feb)/.test(host) ||
      (ipv4 && (
        octets[0] === 10 || octets[0] === 127 ||
        (octets[0] === 169 && octets[1] === 254) ||
        (octets[0] === 172 && octets[1] >= 16 && octets[1] <= 31) ||
        (octets[0] === 192 && octets[1] === 168)
      ))
    ) {
      return 'lan'
    }
    return 'public'
  } catch {
    return 'custom'
  }
}

/**
 * Type guard — is this value shaped like an `ApiEndpoint`? Used by the
 * pairing parser to filter malformed candidates without throwing.
 */
export function isApiEndpointShape(v: unknown): v is ApiEndpoint {
  if (typeof v !== 'object' || v === null) return false
  const o = v as Record<string, unknown>
  return (
    typeof o.host === 'string' &&
    typeof o.port === 'number' &&
    (typeof o.tls === 'boolean' || o.tls === undefined)
  )
}

/**
 * Type guard — is this value shaped like a `RelayEndpoint`? `transport_hint`
 * is the wire name; `transportHint` is the JS-side name. The parser handles
 * the rename.
 */
export function isRelayEndpointShape(v: unknown): v is { url: string; transport_hint?: string } {
  if (typeof v !== 'object' || v === null) return false
  const o = v as Record<string, unknown>
  return (
    typeof o.url === 'string' &&
    (typeof o.transport_hint === 'string' || o.transport_hint === undefined)
  )
}
